slasher/detection/attestations/spanner.go

```diff
@@ -0,0 +1,197 @@
+	"github.com/prysmaticlabs/prysm/shared/params"
+	"go.opencensus.io/trace"
+)
+
+// DetectionKind defines an enum type that
+// gives us information on the type of slashable offense
+// found when analyzing validator min-max spans.
+type DetectionKind int
+
+const (
+	// DoubleVote denotes a slashable offense in which
+	// a validator cast two conflicting attestations within
+	// the same target epoch.
+	DoubleVote DetectionKind = iota
+	// SurroundVote denotes a slashable offense in which
+	// a validator surrounded or was surrounded by a previous
+	// attestation created by the same validator.
+	SurroundVote
+)
+
+// DetectionResult tells us the kind of slashable
+// offense found from detecting on min-max spans +
+// the slashable epoch for the offense.
+type DetectionResult struct {
+	Kind           DetectionKind
+	SlashableEpoch uint64
+}
+
+// SpanDetector defines a struct which can detect slashable
+// attestation offenses by tracking validator min-max
+// spans from validators.
+type SpanDetector struct {
+	// Slice of epochs for valindex => min-max span.
+	spans []map[uint64][2]uint16
+	lock  sync.RWMutex
+}
+
+// NewSpanDetector creates a new instance of a struct tracking
+// several epochs of min-max spans for each validator in
+// the beacon state.
+func NewSpanDetector() *SpanDetector {
+	return &SpanDetector{
+		spans: make([]map[uint64][2]uint16, 256),
+	}
+}
+
+// DetectSlashingForValidator uses a validator index and its corresponding
+// min-max spans during an epoch to detect an epoch in which the validator
+// committed a slashable attestation.
+func (s *SpanDetector) DetectSlashingForValidator(
+	ctx context.Context,
+	validatorIdx uint64,
+	sourceEpoch uint64,
+	targetEpoch uint64,
+) (*DetectionResult, error) {
+	ctx, span := trace.StartSpan(ctx, "detection.DetectSlashingForValidator")
+	defer span.End()
+	if (targetEpoch - sourceEpoch) > params.BeaconConfig().WeakSubjectivityPeriod {
+		return nil, fmt.Errorf(
+			"attestation span was greater than weak subjectivity period %d, received: %d",
+			params.BeaconConfig().WeakSubjectivityPeriod,
+			targetEpoch-sourceEpoch,
+		)
+	}
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+	distance := uint16(targetEpoch - sourceEpoch)
+	numSpans := uint64(len(s.spans))
+	if sp := s.spans[sourceEpoch%numSpans]; sp != nil {
+		minSpan := sp[validatorIdx][0]
+		if minSpan > 0 && minSpan < distance {
+			return &DetectionResult{
+				Kind:           SurroundVote,
+				SlashableEpoch: uint64(minSpan) + sourceEpoch,
+			}, nil
+		}
+
+		maxSpan := sp[validatorIdx][1]
+		if maxSpan > distance {
+			return &DetectionResult{
+				Kind:           SurroundVote,
+				SlashableEpoch: uint64(maxSpan) + sourceEpoch,
+			}, nil
+		}
+	}
+	return nil, nil
+}
+
+// SpansForValidatorByEpoch returns the specific min-max span for a
+// validator index in a given epoch.
+func (s *SpanDetector) SpansForValidatorByEpoch(ctx context.Context, valIdx uint64, epoch uint64) ([2]uint16, error) {
```

Comment:
Change this to `SpanForEpochByValidator`?

Reply:
Done @0xKiwi